readmeio/rdme
# feat: major version tags

## 🧰 Changes

I've always really liked how [`actions/setup-node`](https://github.com/actions/setup-node) and [`actions/checkout` ](https://github.com/actions/checkout) use top-level tags for major versions, which are [recreated every time there's a tag for that major version](https://github.com/actions/setup-node/tags). It makes it a lot easier to point to the latest in a workflow (as well as in the docs):

```yml
steps:
  - uses: actions/checkout@v3
  - uses: actions/setup-node@v3
```

We've been shipping a lot of updates lately (a good thing!), but it's gotta be annoying for people that have Dependabot set up, since they're likely using an outdated version already. Now that we have [a streamlined release process via `npm run bump`](https://github.com/readmeio/rdme/blob/main/MAINTAINERS.md), I figured we should enhance the process a bit so it automatically recreates a major tag with every release.

As part of this:

- [x] I did a bunch of refactoring to try and consolidate our versioning logic as much as possible
- [x] I also [addressed a little deprecation notice with our GitHub Actions outputs](https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/) (more details on this in the comments)
- [x] This major tag will be surfaced both in our docs and in any workflow files created via our onboarding.

## 🧬 QA & Testing

I've done a few dry runs of `npm run bump` and confirmed that the tag is created:

<img width="620" alt="Screen Shot 2022-11-29 at 9 59 56 AM" src="https://user-images.githubusercontent.com/8854718/204598580-f8ce6622-6de8-4431-8c12-f0e508afc704.png">

<img width="358" alt="Screen Shot 2022-11-29 at 11 30 45 AM" src="https://user-images.githubusercontent.com/8854718/204600549-6091014b-6c35-451d-833f-5c02d9268cf3.png">

I also tried with a prerelease tag and confirmed that the major tag is **not** created:

<img width="475" alt="Screen Shot 2022-11-29 at 11 22 26 AM" src="https://user-images.githubusercontent.com/8854718/204598901-7bc4a641-395d-4cf2-8a35-47c493594908.png">

You can follow these steps to try it out yourself:
1. Check out this branch
2. Run something like `npm run bump -- 8.1.2` or `npm run bump -- 8.1.2-beta.0`
3. Inspect the commit created (but do NOT push it) and make sure that everything looks good! Just make sure to delete the tags you create (using `git tag -d [tag]`) and don't push anything to GitHub.
